coreneuron/io/core2nrn_data_return.cpp

```diff
@@ -0,0 +1,174 @@
+ *  src is a contiguous array of sz segments of size stride. The stride
+ *  may be slightly greater than n for purposes of alignment.
+ *  Each of the sz segments of src are permuted.
+ */
+static void soa2aos_inverse_permute_copy(size_t n, int sz, int stride,
+  double* src, double** dest, int* permute)
+{
+  // src is soa and permuted. dest is n pointers to sz doubles (aos).
+  for (size_t instance = 0; instance < n; ++instance) {
+    double* d = dest[instance];
+    double* s = src + permute[instance];
+    for (int i = 0; i < sz; ++i) {
+      d[i] = s[i*stride];
+    }
+  }
+}
+
+/** @brief SoA unpermuted mechanism data copied to unpermuted AoS data.
+ *  dest is an array of n pointers to the beginning of each sz length array.
+ *  src is a contiguous array of sz segments of size stride. The stride
+ *  may be slightly greater than n for purposes of alignment.
+ *  Each of the sz segments of src have the same order as the n pointers
+ *  of dest.
+ */
+static void soa2aos_unpermuted_copy(size_t n, int sz, int stride,
+  double* src, double** dest)
+{
+  // src is soa and permuted. dest is n pointers to sz doubles (aos).
+  for (size_t instance = 0; instance < n; ++instance) {
+    double* d = dest[instance];
+    double* s = src + instance;
+    for (int i = 0; i < sz; ++i) {
+      d[i] = s[i*stride];
+    }
+  }
+}
+
+/** @brief AoS mechanism data copied to AoS data.
+ *  dest is an array of n pointers to the beginning of each sz length array.
+ *  src is a contiguous array of n segments of size sz.
+ */
+static void aos2aos_copy(size_t n, int sz, double* src, double** dest) {
+  for (size_t instance = 0; instance < n; ++instance) {
+    double* d = dest[instance];
+    double* s = src + (instance*sz);
+    std::copy(s, s + sz, d);
+  }
+}
+      
+/** @brief copy data back to NEURON.
+ *  Copies t, voltage, i_membrane_ if it used, and mechanism param data.
+ */
+void core2nrn_data_return() {
+  if (!nrn2core_type_return_) {
+    return;
+  }
+  for (int tid = 0; tid < nrn_nthread; ++tid) {
+    size_t n = 0;
+    double* data = nullptr;
+    double** mdata = nullptr;
+    NrnThread& nt = nrn_threads[tid];
+
+    n = (*nrn2core_type_return_)(0, tid, data, mdata); // 0 means time
+    if (n) { // not the empty thread
+      data[0] = nt._t;
+    }
+
+    if (nt.end) { // transfer voltage and possibly i_membrane_
+      n = (*nrn2core_type_return_)(voltage, tid, data, mdata);
+      assert(n == size_t(nt.end) && data);
+      inverse_permute_copy(n, nt._actual_v, data, nt._permute);
+
+      if (nt.nrn_fast_imem) {
+        n = (*nrn2core_type_return_)(i_membrane_, tid, data, mdata);
+        assert(n == size_t(nt.end) && data);
+        inverse_permute_copy(n, nt.nrn_fast_imem->nrn_sav_rhs, data, nt._permute);
+      }
+    }
+
+    for (NrnThreadMembList* tml = nt.tml; tml; tml = tml->next) {
+      int mtype = tml->index;
+      Memb_list* ml = tml->ml;
+      n = (*nrn2core_type_return_)(mtype, tid, data, mdata);
```

Comment:
I should look at neuron side but curious how `(*nrn2core_type_return_)(....);` works when called multiple times.

Is that first argument `mtype` decides what type of data is it?

Reply:
tid is the thread index, mtype is the mechanism type. A lot of this is in fact thread safe on the neuron side and I hope that the refactoring project will take advantage of it so that all the loops over nrn_threads can be done in parallel.